src/hydro_system.hpp

```diff
@@ -583,6 +573,95 @@ void HydroSystem<problem_t>::FlattenShocks(amrex::MultiFab const &q_mf, amrex::M
 	});
 }
 
+// to ensure that physical quantities are within reasonable
+// floors and ceilings which can be set in the param file
+template <typename problem_t>
+void HydroSystem<problem_t>::EnforceLimits(amrex::Real const densityFloor, amrex::Real const pressureFloor, amrex::Real const speedCeiling,
+					   amrex::Real const tempCeiling, amrex::Real const tempFloor, amrex::MultiFab &state_mf)
+{
+
+	amrex::Real const rho_floor = densityFloor; // workaround nvcc bug
+	amrex::Real const P_floor = pressureFloor;
+	auto state = state_mf.arrays();
+
+	amrex::ParallelFor(state_mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+		amrex::Real const rho = state[bx](i, j, k, density_index);
+		amrex::Real const vx1 = state[bx](i, j, k, x1Momentum_index) / rho;
+		amrex::Real const vx2 = state[bx](i, j, k, x2Momentum_index) / rho;
+		amrex::Real const vx3 = state[bx](i, j, k, x3Momentum_index) / rho;
+		amrex::Real const vsq = (vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real const v_abs = std::sqrt(vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real Etot = state[bx](i, j, k, energy_index);
+		amrex::Real Eint = state[bx](i, j, k, internalEnergy_index);
+		amrex::Real Ekin = rho * vsq / 2.;
+		amrex::Real rho_new = rho;
+
+		if (rho < rho_floor) {
+			rho_new = rho_floor;
+			state[bx](i, j, k, density_index) = rho_new;
+			state[bx](i, j, k, internalEnergy_index) = Eint * rho_new / rho;
+			state[bx](i, j, k, energy_index) = rho_new * vsq / 2. + (Etot - Ekin);
+			if (nscalars_ > 0) {
+				for (int n = 0; n < nscalars_; ++n) {
+					state[bx](i, j, k, scalar0_index + n) *= rho / rho_new;
```

Comment:
One of the division-by-zero errors is here. This happens because `rho_floor` is zero. In this case, I'm not sure what the correct thing to do is. I think the options for this case are either: i) don't change the passive scalars ii) set the passive scalars to zero.

Reply:
I think the correct behavior in this case is to set the passive scalar to zero as well. Leaving it non-zero is an invitation to trouble, because in the common use case where the passive scalar density represents a partial density of some species, we want to avoid a situation where the sum of partial densities is larger than the total density.